controller/recurring_job_controller.go

```diff
@@ -0,0 +1,429 @@
+
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	typedv1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhinformers "github.com/longhorn/longhorn-manager/k8s/pkg/client/informers/externalversions/longhorn/v1beta1"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+)
+
+type RecurringJobController struct {
+	*baseController
+
+	namespace string
+
+	controllerID   string
+	ManagerImage   string
+	serviceAccount string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	ds *datastore.DataStore
+
+	rjStoreSynced cache.InformerSynced
+}
+
+func NewRecurringJobController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	recurringJobInformer lhinformers.RecurringJobInformer,
+	kubeClient clientset.Interface,
+	namespace, controllerID, serviceAccount, managerImage string,
+) *RecurringJobController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&typedv1core.EventSinkImpl{Interface: typedv1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
+
+	rjc := &RecurringJobController{
+		baseController: newBaseController("longhorn-recurring-job", logger),
+
+		namespace:      namespace,
+		controllerID:   controllerID,
+		ManagerImage:   managerImage,
+		serviceAccount: serviceAccount,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: "longhorn-recurring-job-controller"}),
+
+		ds: ds,
+
+		rjStoreSynced: recurringJobInformer.Informer().HasSynced,
+	}
+
+	recurringJobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    rjc.enqueueRecurringJob,
+		UpdateFunc: func(old, cur interface{}) { rjc.enqueueRecurringJob(cur) },
+		DeleteFunc: rjc.enqueueRecurringJob,
+	})
+
+	return rjc
+}
+
+func (control *RecurringJobController) enqueueRecurringJob(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	control.queue.AddRateLimited(key)
+}
+
+func (control *RecurringJobController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer control.queue.ShutDown()
+
+	logrus.Infof("Start Longhorn Recurring Job controller")
```

Comment:
"Staring ....". Encourage use verb+ing/ed in the log.

Reply:
Fixed.